examples/ex10.cpp

```diff
@@ -177,11 +177,7 @@ int main(int argc, char *argv[])
    args.AddOption(&order, "-o", "--order",
                   "Order (degree) of the finite elements.");
    args.AddOption(&ode_solver_type, "-s", "--ode-solver",
-                  "ODE solver: 1 - Backward Euler, 2 - SDIRK2, 3 - SDIRK3,\n\t"
-                  "            11 - Forward Euler, 12 - RK2,\n\t"
-                  "            13 - RK3 SSP, 14 - RK4."
-                  "            22 - Implicit Midpoint Method,\n\t"
-                  "            23 - SDIRK23 (A-stable), 24 - SDIRK34");
+                  ODESolver::Types.c_str());
```

Comment:
It would be cool if we didn't need to call `c_str()` here (i.e. if `AddOption` took `const std::string &` as an argument).

If you're interested, here is a patch that should do it.

<details>
<summary>Click to expand patch</summary>

```diff
diff --git a/general/optparser.cpp b/general/optparser.cpp
index a59854899..45b6f2a24 100644
--- a/general/optparser.cpp
+++ b/general/optparser.cpp
@@ -150,7 +150,7 @@ void parseVector(char * str, Vector & var)
 
 void OptionsParser::Parse()
 {
-   option_check.SetSize(options.Size());
+   option_check.SetSize(options.size());
    option_check = 0;
    for (int i = 1; i < argc; )
    {
@@ -163,7 +163,7 @@ void OptionsParser::Parse()
 
       for (int j = 0; true; j++)
       {
-         if (j >= options.Size())
+         if (j >= options.size())
          {
             // unrecognized option
             error_type = 2;
@@ -171,8 +171,7 @@ void OptionsParser::Parse()
             return;
          }
 
-         if (strcmp(argv[i], options[j].short_name) == 0 ||
-             strcmp(argv[i], options[j].long_name) == 0)
+         if (options[j].short_name == argv[i] || options[j].long_name == argv[i])
          {
             OptionType type = options[j].type;
 
@@ -239,7 +238,7 @@ void OptionsParser::Parse()
    }
 
    // check for missing required options
-   for (int i = 0; i < options.Size(); i++)
+   for (int i = 0; i < options.size(); i++)
       if (options[i].required &&
           (option_check[i] == 0 ||
            (options[i].type == ENABLE && option_check[++i] == 0)))
@@ -333,7 +332,7 @@ void OptionsParser::PrintOptions(ostream &os) const
    static const char *indent = "   ";
 
    os << "Options used:\n";
-   for (int j = 0; j < options.Size(); j++)
+   for (int j = 0; j < options.size(); j++)
    {
       OptionType type = options[j].type;
 
@@ -415,7 +414,7 @@ void OptionsParser::PrintHelp(ostream &os) const
 
    os << indent << "-h" << seprtr << "--help" << descr_sep
       << "Print this help message and exit.\n" << line_sep;
-   for (int j = 0; j < options.Size(); j++)
+   for (int j = 0; j < options.size(); j++)
    {
       OptionType type = options[j].type;
 
@@ -451,7 +450,7 @@ void OptionsParser::PrintHelp(ostream &os) const
       }
       os << descr_sep;
 
-      if (options[j].description)
+      if (!options[j].description.empty())
       {
          os << options[j].description << '\n';
       }
diff --git a/general/optparser.hpp b/general/optparser.hpp
index eecf0f74a..197a1c9a1 100644
--- a/general/optparser.hpp
+++ b/general/optparser.hpp
@@ -14,6 +14,8 @@
 
 #include "../config/config.hpp"
 #include "array.hpp"
+#include <string>
+#include <vector>
 
 namespace mfem
 {
@@ -38,22 +40,22 @@ private:
    {
       OptionType type;
       void *var_ptr;
-      const char *short_name;
-      const char *long_name;
-      const char *description;
+      std::string short_name;
+      std::string long_name;
+      std::string description;
       bool required;
 
       Option() = default;
 
-      Option(OptionType type_, void *var_ptr_, const char *short_name_,
-             const char *long_name_, const char *description_, bool req)
+      Option(OptionType type_, void *var_ptr_, const std::string &short_name_,
+             const std::string &long_name_, const std::string &description_, bool req)
          : type(type_), var_ptr(var_ptr_), short_name(short_name_),
            long_name(long_name_), description(description_), required(req) { }
    };
 
    int argc;
    char **argv;
-   Array<Option> options;
+   std::vector<Option> options;
    Array<int> option_check;
    // error_type can be:
    //  0 - no error
@@ -79,69 +81,66 @@ public:
    /** @brief Add a boolean option and set 'var' to receive the value.
        Enable/disable tags are used to set the bool to true/false
        respectively. */
-   void AddOption(bool *var, const char *enable_short_name,
-                  const char *enable_long_name, const char *disable_short_name,
-                  const char *disable_long_name, const char *description,
+   void AddOption(bool *var, const std::string &enable_short_name,
+                  const std::string &enable_long_name, const std::string &disable_short_name,
+                  const std::string &disable_long_name, const std::string &description,
                   bool required = false)
    {
-      options.Append(Option(ENABLE, var, enable_short_name, enable_long_name,
-                            description, required));
-      options.Append(Option(DISABLE, var, disable_short_name, disable_long_name,
-                            description, required));
+      options.emplace_back(ENABLE, var, enable_short_name, enable_long_name,
+                           description, required);
+      options.emplace_back(DISABLE, var, disable_short_name, disable_long_name,
+                           description, required);
    }
 
    /// Add an integer option and set 'var' to receive the value.
-   void AddOption(int *var, const char *short_name, const char *long_name,
-                  const char *description, bool required = false)
+   void AddOption(int *var, const std::string &short_name,
+                  const std::string &long_name,
+                  const std::string &description, bool required = false)
    {
-      options.Append(Option(INT, var, short_name, long_name, description,
-                            required));
+      options.emplace_back(INT, var, short_name, long_name, description, required);
    }
 
    /// Add a double option and set 'var' to receive the value.
-   void AddOption(double *var, const char *short_name, const char *long_name,
-                  const char *description, bool required = false)
+   void AddOption(double *var, const std::string &short_name,
+                  const std::string &long_name,
+                  const std::string &description, bool required = false)
    {
-      options.Append(Option(DOUBLE, var, short_name, long_name, description,
-                            required));
+      options.emplace_back(DOUBLE, var, short_name, long_name, description, required);
    }
 
    /// Add a string (char*) option and set 'var' to receive the value.
-   void AddOption(const char **var, const char *short_name,
-                  const char *long_name, const char *description,
+   void AddOption(const char **var, const std::string &short_name,
+                  const std::string &long_name, const std::string &description,
                   bool required = false)
    {
-      options.Append(Option(STRING, var, short_name, long_name, description,
-                            required));
+      options.emplace_back(STRING, var, short_name, long_name, description, required);
    }
 
    /// Add a string (std::string) option and set 'var' to receive the value.
-   void AddOption(std::string *var, const char *short_name,
-                  const char *long_name, const char *description,
+   void AddOption(std::string *var, const std::string &short_name,
+                  const std::string &long_name, const std::string &description,
                   bool required = false)
    {
-      options.Append(Option(STD_STRING, var, short_name, long_name, description,
-                            required));
+      options.emplace_back(STD_STRING, var, short_name, long_name, description,
+                           required);
    }
 
    /** Add an integer array (separated by spaces) option and set 'var' to
        receive the values. */
-   void AddOption(Array<int> * var, const char *short_name,
-                  const char *long_name, const char *description,
+   void AddOption(Array<int> * var, const std::string &short_name,
+                  const std::string &long_name, const std::string &description,
                   bool required = false)
    {
-      options.Append(Option(ARRAY, var, short_name, long_name, description,
-                            required));
+      options.emplace_back(ARRAY, var, short_name, long_name, description, required);
    }
 
    /** Add a vector (doubles separated by spaces) option and set 'var' to
        receive the values. */
-   void AddOption(Vector * var, const char *short_name,
-                  const char *long_name, const char *description,
+   void AddOption(Vector * var, const std::string &short_name,
+                  const std::string &long_name, const std::string &description,
                   bool required = false)
    {
-      options.Append(Option(VECTOR, var, short_name, long_name, description,
-                            required));
+      options.emplace_back(VECTOR, var, short_name, long_name, description, required);
    }
 
    /** @brief Parse the command-line options.
```

</details>

Reply:
I agree that the `c_str()` call is clumsy. I can also see that `std::string` might be preferred to `char *`.
But perhaps this should be a separate PR? With some discussion perhaps? Or is there consensus/precedent to move from `C` syntax to `C++`?

Reply:
Disadvantage of doing it in this PR: it's combining different sets of changes that are only tangentially related.
Advantage of doing it in this PR: we're already discussing it now, less overhead in the review process.

Thoughts on this, @v-dobrev?

Reply:
I personally think using `.c_str()` is fine here. 🦖 😄

That said, I'm not against changing the `OptionsParser` class to take `const std::string &` inputs instead of `const char *` but let's do it in a separate PR.

Reply:
Handled in separate PR #3810

Reply:
Just to clarify: #3810 is not what @pazner was suggesting here -- he's suggesting replacing the options-names and documentation-string arguments of the various `AddOption` methods in `mfem::OptionsParser` from `const char *` to `const std::string &`. That will be a new PR -- basically the patch that he posted above.